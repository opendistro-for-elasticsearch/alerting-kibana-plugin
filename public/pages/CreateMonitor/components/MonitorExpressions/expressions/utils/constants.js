/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export const POPOVER_STYLE = { zIndex: '200' };
export const EXPRESSION_STYLE = { padding: '20px', whiteSpace: 'nowrap' };
export const Expressions = {
  THRESHOLD: 'THRESHOLD',
  WHEN: 'WHEN',
  OF_FIELD: 'OF_FIELD',
  OVER: 'OVER',
  FOR_THE_LAST: 'FOR_THE_LAST',
  WHERE: 'WHERE',
};
export const NUMBER_TYPES = [
  'long',
  'integer',
  'short',
  'byte',
  'double',
  'float',
  'half_float',
  'scaled_float',
];
export const UNITS_OF_TIME = [
  { value: 'm', text: 'minute(s)' },
  { value: 'h', text: 'hour(s)' },
  { value: 'd', text: 'day(s)' },
];

export const WHERE_BOOLEAN_FILTERS = [
  { text: 'Select value', value: '' },
  { text: 'True', value: true },
  { text: 'False', value: false },
];

export const OPERATORS_MAP = {
  IS: 'is',
  IS_NOT: 'is_not',
  IS_NULL: 'is_null',
  IS_NOT_NULL: 'is_not_null',
  IS_GREATER: 'is_greater',
  IS_GREATER_EQUAL: 'is_greater_equal',
  IS_LESS: 'is_lesser',
  IS_LESS_EQUAL: 'is_lesser_equal',
  IS_LESS_EQUAL: 'is_lesser_equal',
  STARTS_WITH: 'starts_with',
  ENDS_WITH: 'ends_with',
  CONTAINS: 'contains',
  NOT_CONTAINS: 'does_not_contains',
  IN_RANGE: 'in_range',
  NOT_IN_RANGE: 'not_in_range',
};

export const COMPARISON_OPERATORS = [
  { text: 'is', value: OPERATORS_MAP.IS, dataTypes: ['number', 'text', 'keyword', 'boolean'] },
  {
    text: 'is not',
    value: OPERATORS_MAP.IS_NOT,
    dataTypes: ['number', 'text', 'keyword', 'boolean'],
  },
  {
    text: 'is null',
    value: OPERATORS_MAP.IS_NULL,
    dataTypes: ['number', 'text', 'keyword', 'boolean'],
  },
  {
    text: 'is not null',
    value: OPERATORS_MAP.IS_NOT_NULL,
    dataTypes: ['number', 'text', 'keyword'],
  },
  { text: 'is greater than', value: OPERATORS_MAP.IS_GREATER, dataTypes: ['number'] },
  { text: 'is greater than equal', value: OPERATORS_MAP.IS_GREATER_EQUAL, dataTypes: ['number'] },
  { text: 'is less than', value: OPERATORS_MAP.IS_LESS, dataTypes: ['number'] },
  { text: 'is less than equal', value: OPERATORS_MAP.IS_LESS_EQUAL, dataTypes: ['number'] },
  { text: 'is in range', value: OPERATORS_MAP.IN_RANGE, dataTypes: ['number'] },
  { text: 'is not in range', value: OPERATORS_MAP.NOT_IN_RANGE, dataTypes: ['number'] },
  { text: 'starts with', value: OPERATORS_MAP.STARTS_WITH, dataTypes: ['text', 'keyword'] },
  { text: 'ends with', value: OPERATORS_MAP.ENDS_WITH, dataTypes: ['text', 'keyword'] },
  { text: 'contains', value: OPERATORS_MAP.CONTAINS, dataTypes: ['text', 'keyword'] },
  { text: 'does not contains', value: OPERATORS_MAP.NOT_CONTAINS, dataTypes: ['text'] },
];
export const OVER_TYPES = [{ value: 'all documents', text: 'all documents' }];

export const AGGREGATION_TYPES = [
  { value: 'avg', text: 'average()' },
  { value: 'count', text: 'count()' },
  { value: 'sum', text: 'sum()' },
  { value: 'min', text: 'min()' },
  { value: 'max', text: 'max()' },
];
