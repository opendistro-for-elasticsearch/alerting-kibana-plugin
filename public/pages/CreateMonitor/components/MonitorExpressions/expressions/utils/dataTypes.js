/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { NUMBER_TYPES } from './constants';

export function getFieldsForType(dataTypes, type) {
  if (type === 'number') {
    return NUMBER_TYPES.reduce(
      (options, type) => options.concat(getFieldsForType(dataTypes, type)),
      []
    );
  }
  if (dataTypes[type]) {
    return [...dataTypes[type].values()];
  }
  return [];
}

export const getIndexFields = (dataTypes, allowedTypes) =>
  allowedTypes.map((type) => ({
    label: type,
    options: getFieldsForType(dataTypes, type).map((field) => ({ label: field, type })),
  }));

export const getFilteredIndexFields = (dataTypes, allowedTypes, fieldsToInclude) =>
  allowedTypes.map((type) => ({
    label: type,
    options: getFieldsForType(dataTypes, type)
      .filter((field) => fieldsToInclude.includes(field))
      .map((field) => ({ label: field, type })),
  }));
