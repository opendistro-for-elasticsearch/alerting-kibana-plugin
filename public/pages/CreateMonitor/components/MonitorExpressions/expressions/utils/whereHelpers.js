/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import _ from 'lodash';
import { COMPARISON_OPERATORS, OPERATORS_MAP } from './constants';
import { TRIGGER_COMPARISON_OPERATORS } from '../../../../../CreateTrigger/containers/DefineAggregationTrigger/DefineAggregationTrigger';

export const getOperators = (fieldType) =>
  COMPARISON_OPERATORS.reduce(
    (acc, currentOperator) =>
      currentOperator.dataTypes.includes(fieldType)
        ? [...acc, { text: currentOperator.text, value: currentOperator.value }]
        : acc,
    []
  );

export const isRangeOperator = (selectedOperator) =>
  [OPERATORS_MAP.IN_RANGE, OPERATORS_MAP.NOT_IN_RANGE].includes(selectedOperator);
export const isNullOperator = (selectedOperator) =>
  [OPERATORS_MAP.IS_NULL, OPERATORS_MAP.IS_NOT_NULL].includes(selectedOperator);

export const displayText = (whereValues) => {
  const comparisonOperators = _.concat(COMPARISON_OPERATORS, TRIGGER_COMPARISON_OPERATORS);

  const whereFieldName = _.get(whereValues, 'fieldName[0].label', undefined);
  if (!whereFieldName) {
    return 'all fields are included';
  }
  const selectedOperator = _.get(whereValues, 'operator', 'is');
  const operatorObj =
    comparisonOperators.find((operator) => operator.value === selectedOperator) || {};
  const initialText = `${whereFieldName} ${operatorObj.text || ''}`;

  if (isRangeOperator(selectedOperator)) {
    const startRange = _.get(whereValues, 'fieldRangeStart', 0);
    const endRange = _.get(whereValues, 'fieldRangeEnd', 0);
    return `${initialText} from ${startRange} to ${endRange}`;
  } else if (isNullOperator(selectedOperator)) {
    return `${initialText}`;
  } else {
    const value = _.get(whereValues, 'fieldValue', '');
    return `${initialText} ${value}`;
  }
};

export const validateRange = (value, whereFilters) => {
  if (value === '') return 'Required';
  if (whereFilters.fieldRangeEnd < value) {
    return 'Start should be less than end range';
  }
  if (value < whereFilters.fieldRangeStart) {
    return 'End should be greater than start range';
  }
};
