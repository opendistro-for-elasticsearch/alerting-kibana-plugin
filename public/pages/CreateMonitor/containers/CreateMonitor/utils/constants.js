import { OPERATORS_MAP } from '../../../components/MonitorExpressions/expressions/utils/constants';

/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export const BUCKET_COUNT = 5;

export const MATCH_ALL_QUERY = JSON.stringify({ size: 0, query: { match_all: {} } }, null, 4);

export const FORMIK_INITIAL_VALUES = {
  /* CONFIGURE MONITOR */
  name: '',
  disabled: false,
  frequency: 'interval',
  timezone: [],
  daily: 0,
  period: { interval: 1, unit: 'MINUTES' },
  weekly: { mon: false, tue: false, wed: false, thur: false, fri: false, sat: false, sun: false },
  monthly: { type: 'day', day: 1 },
  cronExpression: '0 */1 * * *',

  /* DEFINE MONITOR */
  searchType: 'graph',
  index: [],
  timeField: '',
  query: MATCH_ALL_QUERY,
  aggregationType: 'count',
  fieldName: [],
  overDocuments: 'all documents',
  groupedOverTop: 5,
  groupedOverFieldName: 'bytes',
  bucketValue: 1,
  bucketUnitOfTime: 'h', // m = minute, h = hour, d = day
  where: {
    fieldName: [],
    operator: OPERATORS_MAP.IS,
    fieldValue: '',
    fieldRangeStart: 0,
    fieldRangeEnd: 0,
  },
  detectorId: '',
};
