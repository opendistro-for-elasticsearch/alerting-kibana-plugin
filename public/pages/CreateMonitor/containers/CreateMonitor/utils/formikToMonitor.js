/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import _ from 'lodash';
import moment from 'moment-timezone';
import { BUCKET_COUNT } from './constants';
import { SEARCH_TYPE } from '../../../../../utils/constants';
import { OPERATORS_QUERY_MAP } from './whereFilters';

export function formikToMonitor(values) {
  const uiSchedule = formikToUiSchedule(values);
  const schedule = buildSchedule(values.frequency, uiSchedule);
  const uiSearch = formikToUiSearch(values);
  return {
    name: values.name,
    type: 'monitor',
    enabled: !values.disabled,
    schedule,
    inputs: [formikToSearch(values)],
    triggers: [],
    ui_metadata: {
      schedule: uiSchedule,
      search: uiSearch,
    },
  };
}

export function formikToInputs(values) {
  const isAD = values.searchType === SEARCH_TYPE.AD;
  return [isAD ? formikToAd(values) : formikToSearch(values)];
}

export function formikToSearch(values) {
  const isAD = values.searchType === SEARCH_TYPE.AD;
  let query = isAD ? formikToAdQuery(values) : formikToQuery(values);
  const indices = isAD ? ['.opendistro-anomaly-results*'] : formikToIndices(values);

  return {
    search: {
      indices,
      query,
    },
  };
}

export function formikToAdQuery(values) {
  return {
    size: 1,
    sort: [{ anomaly_grade: 'desc' }, { confidence: 'desc' }],
    query: {
      bool: {
        filter: [
          {
            range: {
              data_end_time: {
                from: '{{period_end}}||-' + values.period.interval + 'm',
                to: '{{period_end}}',
                include_lower: true,
                include_upper: true,
              },
            },
          },
          {
            term: {
              detector_id: {
                value: values.detectorId,
              },
            },
          },
        ],
      },
    },
    aggregations: {
      max_anomaly_grade: {
        max: {
          field: 'anomaly_grade',
        },
      },
    },
  };
}

export function formikToAd(values) {
  return {
    anomaly_detector: {
      detector_id: values.detectorId,
    },
  };
}
export function formikToUiSearch(values) {
  const {
    searchType,
    aggregationType,
    timeField,
    fieldName: [{ label: fieldName = '' } = {}],
    overDocuments,
    groupedOverTop,
    groupedOverFieldName,
    bucketValue,
    bucketUnitOfTime,
    where,
  } = values;
  return {
    searchType,
    aggregationType,
    timeField,
    fieldName,
    overDocuments,
    groupedOverTop,
    groupedOverFieldName,
    bucketValue,
    bucketUnitOfTime,
    where,
  };
}

export function formikToIndices(values) {
  return values.index.map(({ label }) => label);
}

export function formikToQuery(values) {
  const isGraph = values.searchType === SEARCH_TYPE.GRAPH;
  return isGraph ? formikToGraphQuery(values) : formikToExtractionQuery(values);
}

export function formikToExtractionQuery(values) {
  return JSON.parse(values.query);
}

export function formikToGraphQuery(values) {
  const { bucketValue, bucketUnitOfTime } = values;
  const whenAggregation = formikToWhenAggregation(values);
  const timeField = values.timeField;
  const filters = [
    {
      range: {
        [timeField]: {
          gte: `{{period_end}}||-${Math.round(bucketValue)}${bucketUnitOfTime}`,
          lte: '{{period_end}}',
          format: 'epoch_millis',
        },
      },
    },
  ];
  const whereClause = formikToWhereClause(values);
  if (whereClause) {
    filters.push({ ...whereClause });
  }
  return {
    size: 0,
    aggregations: whenAggregation,
    query: {
      bool: {
        filter: filters,
      },
    },
  };
}

export function formikToUiGraphQuery(values) {
  const { bucketValue, bucketUnitOfTime } = values;
  const overAggregation = formikToUiOverAggregation(values);
  const timeField = values.timeField;
  const filters = [
    {
      range: {
        [timeField]: {
          // default range window to [BUCKET_COUNT] * the date histogram interval
          gte: `now-${bucketValue * BUCKET_COUNT}${bucketUnitOfTime}`,
          lte: 'now',
        },
      },
    },
  ];
  const whereClause = formikToWhereClause(values);
  if (whereClause) {
    filters.push({ ...whereClause });
  }
  return {
    size: 0,
    aggregations: overAggregation,
    query: {
      bool: {
        filter: filters,
      },
    },
  };
}

export function formikToUiOverAggregation(values) {
  const whenAggregation = formikToWhenAggregation(values);
  const { bucketValue, bucketUnitOfTime } = values;
  const timeField = values.timeField;

  return {
    over: {
      date_histogram: {
        field: timeField,
        interval: `${bucketValue}${bucketUnitOfTime}`,
        time_zone: moment.tz.guess(),
        min_doc_count: 0,
        extended_bounds: {
          min: `now-${bucketValue * BUCKET_COUNT}${bucketUnitOfTime}`,
          max: 'now',
        },
      },
      aggregations: whenAggregation,
    },
  };
}

export function formikToWhereClause({ where }) {
  if (where.fieldName.length > 0) {
    return OPERATORS_QUERY_MAP[where.operator].query(where);
  }
}

export function formikToWhenAggregation(values) {
  const {
    aggregationType,
    fieldName: [{ label: field } = {}],
  } = values;
  if (aggregationType === 'count' || !field) return {};
  return { when: { [aggregationType]: { field } } };
}

export function formikToUiSchedule(values) {
  return {
    timezone: _.get(values, 'timezone[0].label', null),
    frequency: values.frequency,
    period: values.period,
    daily: values.daily,
    weekly: values.weekly,
    monthly: values.monthly,
    cronExpression: values.cronExpression,
  };
}

export function buildSchedule(scheduleType, values) {
  const {
    period,
    daily,
    weekly,
    monthly: { type, day },
    cronExpression,
    timezone,
  } = values;
  switch (scheduleType) {
    case 'interval': {
      return { period };
    }
    case 'daily': {
      return { cron: { expression: `0 ${daily} * * *`, timezone } };
    }
    case 'weekly': {
      const daysOfWeek = Object.entries(weekly)
        .filter(([day, checked]) => checked)
        .map(([day]) => day.toUpperCase())
        .join(',');
      return { cron: { expression: `0 ${daily} * * ${daysOfWeek}`, timezone } };
    }
    case 'monthly': {
      let dayOfMonth = '?';
      if (type === 'day') {
        dayOfMonth = day;
      }
      return { cron: { expression: `0 ${daily} ${dayOfMonth} */1 *`, timezone } };
    }
    case 'cronExpression':
      return { cron: { expression: cronExpression, timezone } };
  }
}
