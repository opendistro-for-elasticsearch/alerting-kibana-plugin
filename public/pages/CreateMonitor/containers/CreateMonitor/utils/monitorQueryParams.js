import { SEARCH_TYPE } from '../../../../../utils/constants';

export const initializeFromQueryParams = queryParams => {
  return {
    searchType: queryParams.searchType || undefined,
    name: queryParams.name ? `${queryParams.name}-Monitor` : undefined,
    detectorId: queryParams.adId || undefined,
    period:
      queryParams.interval && queryParams.unit
        ? { interval: parseInt(queryParams.interval), unit: queryParams.unit }
        : undefined,
  };
};
