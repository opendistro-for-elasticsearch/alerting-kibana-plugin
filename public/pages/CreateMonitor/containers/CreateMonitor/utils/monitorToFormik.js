/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import _ from 'lodash';
import { FORMIK_INITIAL_VALUES } from './constants';
import { SEARCH_TYPE, INPUTS_DETECTOR_ID } from '../../../../../utils/constants';

// Convert Monitor JSON to Formik values used in UI forms
export default function monitorToFormik(monitor) {
  const formikValues = _.cloneDeep(FORMIK_INITIAL_VALUES);
  if (!monitor) return formikValues;
  const {
    name,
    enabled,
    schedule: { cron: { expression: cronExpression = formikValues.cronExpression, timezone } = {} },
    inputs,
    ui_metadata: { schedule = {}, search = {} } = {},
  } = monitor;
  // Default searchType to query, because if there is no ui_metadata or search then it was created through API or overwritten by API
  // In that case we don't want to guess on the UI what selections a user made, so we will default to just showing the extraction query
  const { searchType = 'query', fieldName } = search;
  const isAD = searchType === SEARCH_TYPE.AD;

  return {
    /* INITIALIZE WITH DEFAULTS */
    ...formikValues,

    /* CONFIGURE MONITOR */
    name,
    disabled: !enabled,

    /* This will overwrite the fields in use by Monitor from ui_metadata */
    ...schedule,
    cronExpression,

    /* DEFINE MONITOR */
    ...search,
    searchType,
    fieldName: fieldName ? [{ label: fieldName }] : [],
    timezone: timezone ? [{ label: timezone }] : [],

    detectorId: isAD //&& _.get(inputs, '0.search.indices.0') === '.opendistro-anomaly-results*'
      ? _.get(inputs, INPUTS_DETECTOR_ID)
      : undefined,
    index: inputs[0].search.indices.map(index => ({ label: index })),
    query: JSON.stringify(inputs[0].search.query, null, 4),
  };
}
