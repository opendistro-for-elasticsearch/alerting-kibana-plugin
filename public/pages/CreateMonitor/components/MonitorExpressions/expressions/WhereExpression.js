/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'formik';
import {
  EuiFlexGroup,
  EuiFlexItem,
  EuiPopover,
  EuiButtonEmpty,
  EuiText,
  EuiSpacer,
  EuiBadge,
} from '@elastic/eui';
import _ from 'lodash';
import {
  Expressions,
  POPOVER_STYLE,
  EXPRESSION_STYLE,
  WHERE_BOOLEAN_FILTERS,
} from './utils/constants';
import {
  getOperators,
  displayText,
  validateRange,
  isNullOperator,
  isRangeOperator,
} from './utils/whereHelpers';
import { isInvalid, required } from '../../../../../utils/validate';
import {
  FormikComboBox,
  FormikSelect,
  FormikFieldNumber,
  FormikFieldText,
} from '../../../../../components/FormControls';
import { getFilteredIndexFields, getIndexFields } from './utils/dataTypes';
import { FORMIK_INITIAL_VALUES } from '../../../containers/CreateMonitor/utils/constants';
import { DATA_TYPES } from '../../../../../utils/constants';
import {
  TRIGGER_COMPARISON_OPERATORS,
  TRIGGER_OPERATORS_MAP,
} from '../../../../CreateTrigger/containers/DefineAggregationTrigger/DefineAggregationTrigger';

const propTypes = {
  formik: PropTypes.object.isRequired,
  dataTypes: PropTypes.object.isRequired,
  onMadeChanges: PropTypes.func.isRequired,
  openedStates: PropTypes.object.isRequired,
  openExpression: PropTypes.func.isRequired,
};

const ALLOWED_TYPES = ['number', 'text', 'keyword', 'boolean'];

class WhereExpression extends Component {
  constructor(props) {
    super(props);
    this.state = {
      fieldPath: !_.isEmpty(`${props.fieldPath}`) ? `${props.fieldPath}.` : '',
    };
  }

  handleFieldChange = (option, field, form) => {
    this.props.onMadeChanges();
    this.resetValues();
    form.setFieldValue(field.name, option);
    // User can remove where condition
    if (option.length === 0) {
      this.resetValues();
      form.setFieldError('where', undefined);
    }
  };

  handleOperatorChange = (e, field) => {
    this.props.onMadeChanges();
    field.onChange(e);
  };

  handleChangeWrapper = (e, field) => {
    this.props.onMadeChanges();
    field.onChange(e);
  };

  handleClosePopOver = async () => {
    const {
      formik: { values },
      closeExpression,
    } = this.props;
    // Explicitly invoking validation, this component unmount after it closes.
    const fieldName = _.get(values, `${this.state.fieldPath}where.fieldName`, '');
    if (fieldName > 0) {
      await this.props.formik.validateForm();
    }
    closeExpression(Expressions.WHERE);
  };

  resetValues = () => {
    const { formik } = this.props;
    formik.setValues({
      ...formik.values,
      where: { ...FORMIK_INITIAL_VALUES.where },
    });
  };

  renderBetweenAnd = () => {
    const {
      formik: { values },
    } = this.props;
    return (
      <EuiFlexGroup alignItems="center">
        <EuiFlexItem>
          <FormikFieldNumber
            name={`${this.state.fieldPath}where.fieldRangeStart`}
            fieldProps={{
              validate: (value) =>
                validateRange(value, _.get(values, `${this.state.fieldPath}where`)),
            }}
            inputProps={{ onChange: this.handleChangeWrapper, isInvalid }}
          />
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiText textAlign="center">TO</EuiText>
        </EuiFlexItem>
        <EuiFlexItem>
          <FormikFieldNumber
            name={`${this.state.fieldPath}where.fieldRangeEnd`}
            fieldProps={{
              validate: (value) =>
                validateRange(value, _.get(values, `${this.state.fieldPath}where`)),
            }}
            inputProps={{ onChange: this.handleChangeWrapper, isInvalid }}
          />
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  renderValueField = (fieldType, fieldOperator) => {
    if (fieldType == DATA_TYPES.NUMBER) {
      return isRangeOperator(fieldOperator) ? (
        this.renderBetweenAnd()
      ) : (
        <FormikFieldNumber
          name={`${this.state.fieldPath}where.fieldValue`}
          fieldProps={{ validate: required }}
          inputProps={{ onChange: this.handleChangeWrapper, isInvalid }}
        />
      );
    } else if (fieldType == DATA_TYPES.BOOLEAN) {
      return (
        <FormikSelect
          name={`${this.state.fieldPath}where.fieldValue`}
          fieldProps={{ validate: required }}
          inputProps={{
            onChange: this.handleChangeWrapper,
            options: WHERE_BOOLEAN_FILTERS,
            isInvalid,
          }}
        />
      );
    } else {
      return (
        <FormikFieldText
          name={`${this.state.fieldPath}where.fieldValue`}
          fieldProps={{ validate: required }}
          inputProps={{ onChange: this.handleChangeWrapper, isInvalid }}
        />
      );
    }
  };

  render() {
    const {
      formik: { values },
      openedStates,
      openExpression,
      dataTypes,
      indexFieldFilters,
      useTriggerFieldOperators,
    } = this.props;
    const { fieldPath } = this.state;
    const indexFields =
      indexFieldFilters !== undefined
        ? getFilteredIndexFields(dataTypes, ALLOWED_TYPES, indexFieldFilters)
        : getIndexFields(dataTypes, ALLOWED_TYPES);
    const fieldType = _.get(values, `${fieldPath}where.fieldName[0].type`, 'number');
    let fieldOperator = _.get(values, `${fieldPath}where.operator`, 'is');
    if (useTriggerFieldOperators && !_.includes(_.values(TRIGGER_OPERATORS_MAP), fieldOperator)) {
      fieldOperator = TRIGGER_OPERATORS_MAP.INCLUDE;
      _.set(values, `${fieldPath}where.operator`, fieldOperator);
    }

    const fieldOperators = useTriggerFieldOperators
      ? TRIGGER_COMPARISON_OPERATORS
      : getOperators(fieldType);

    return (
      <div>
        <EuiText size="xs">
          <h4>Where</h4>
        </EuiText>
        <EuiSpacer size="s" />
        <EuiBadge
          iconSide="right"
          iconType="cross"
          iconOnClick={() => this.resetValues()}
          iconOnClickAriaLabel="Remove where filter"
          onClick={() => {
            openExpression(Expressions.WHERE);
          }}
          onClickAriaLabel="Edit where filter"
        >
          {displayText(_.get(values, `${fieldPath}where`))}
        </EuiBadge>
        <EuiPopover
          id="where-popover"
          button={
            <EuiButtonEmpty
              size="xs"
              data-test-subj="addFilterButton"
              onClick={() => openExpression(Expressions.WHERE)}
            >
              Edit where filter
            </EuiButtonEmpty>
          }
          isOpen={openedStates.WHERE}
          closePopover={this.handleClosePopOver}
          panelPaddingSize="none"
          ownFocus
          withTitle
          anchorPosition="downLeft"
        >
          <div style={POPOVER_STYLE}>
            <EuiFlexGroup style={{ ...EXPRESSION_STYLE }}>
              <EuiFlexItem grow={false} style={{ width: 200 }}>
                <FormikComboBox
                  name={`${fieldPath}where.fieldName`}
                  inputProps={{
                    placeholder: 'Select a field',
                    options: indexFields,
                    onChange: this.handleFieldChange,
                    isClearable: false,
                    singleSelection: { asPlainText: true },
                  }}
                />
              </EuiFlexItem>
              <EuiFlexItem grow={false}>
                <FormikSelect
                  name={`${fieldPath}where.operator`}
                  inputProps={{
                    onChange: this.handleOperatorChange,
                    options: fieldOperators,
                  }}
                />
              </EuiFlexItem>
              {!isNullOperator(fieldOperator) && (
                <EuiFlexItem>{this.renderValueField(fieldType, fieldOperator)}</EuiFlexItem>
              )}
            </EuiFlexGroup>
          </div>
        </EuiPopover>
      </div>
    );
  }
}

WhereExpression.propTypes = propTypes;

export default connect(WhereExpression);
