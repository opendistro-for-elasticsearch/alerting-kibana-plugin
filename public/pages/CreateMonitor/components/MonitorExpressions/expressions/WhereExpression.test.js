/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { Formik } from 'formik';
import { render, mount } from 'enzyme';
import { EuiExpression } from '@elastic/eui';

import { FORMIK_INITIAL_VALUES } from '../../../containers/CreateMonitor/utils/constants';
import WhereExpression from './WhereExpression';
import { FormikFieldText, FormikFieldNumber } from '../../../../../components/FormControls';
import { OPERATORS_MAP } from './utils/constants';

const dataTypes = {
  integer: new Set(['age']),
  text: new Set(['cityName']),
  keyword: new Set(['cityName.keyword']),
};
const openExpression = jest.fn();
const closeExpression = jest.fn();
const getMountWrapper = (state = false) => (
  <Formik
    initialValues={FORMIK_INITIAL_VALUES}
    render={(props) => (
      <WhereExpression
        formik={props}
        dataTypes={dataTypes}
        openedStates={{ WHERE: state }}
        openExpression={openExpression}
        closeExpression={closeExpression}
        onMadeChanges={jest.fn()}
      />
    )}
  />
);

describe('WhereExpression', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  test('renders', () => {
    expect(render(getMountWrapper())).toMatchSnapshot();
  });
  test('calls openExpression when clicking expression', () => {
    const wrapper = mount(getMountWrapper());
    const button = wrapper.find(EuiExpression);
    button.simulate('click');
    wrapper.update();
    expect(openExpression).toHaveBeenCalled();
  });

  test('calls closeExpression when closing popover', () => {
    const wrapper = mount(getMountWrapper(true));
    const button = wrapper.find(EuiExpression);
    button.simulate('keyDown', { keyCode: 27 });
    expect(closeExpression).toHaveBeenCalled();
  });
  test('should render text input for the text data types', async () => {
    const wrapper = mount(getMountWrapper(true));
    wrapper
      .find('[data-test-subj="comboBoxSearchInput"]')
      .hostNodes()
      .simulate('change', { target: { value: 'cityName' } })
      .simulate('keyDown', { key: 'ArrowDown' })
      .simulate('keyDown', { key: 'Enter' })
      .simulate('blur');

    wrapper.update();
    const values = wrapper.find(WhereExpression).props().formik.values;
    expect(values.where.fieldName).toEqual([{ label: 'cityName', type: 'text' }]);
    expect(values.where.operator).toEqual(OPERATORS_MAP.IS);
    expect(wrapper.find(FormikFieldText).length).toBe(1);
    expect(wrapper.find(FormikFieldNumber).length).toBe(0);
  });
});
