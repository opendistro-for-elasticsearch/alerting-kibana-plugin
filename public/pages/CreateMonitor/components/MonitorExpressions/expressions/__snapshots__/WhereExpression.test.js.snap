// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WhereExpression renders 1`] = `
<div
  class="euiPopover euiPopover--anchorDownLeft euiPopover--withTitle"
  id="where-popover"
>
  <div
    class="euiPopover__anchor"
  >
    <button
      class="euiExpression euiExpression-isClickable euiExpression-isUppercase euiExpression--secondary"
    >
      <span
        class="euiExpression__description"
      >
        where
      </span>
       
      <span
        class="euiExpression__value"
      >
        all fields are included
      </span>
    </button>
  </div>
</div>
`;
